Additional info and clarification in listener docs (#1554)
hyperledger/firefly: doc-site/docs/reference/types/_includes/contractlistener_description.md

### Multiple filters

From v1.3.1 onwards, a contract listener can be created with multiple filters under a single topic when supported by the connector. Each filter contains:

- a reference to a specific blockchain event to listen for
- (optional) a specific location/address to listen from
- a connector-specific signature (generated from the event and the location)

> In addition to this list of multiple filters, the listener specifies a single `topic` to identify the stream of events. This new feature will allow better management of contract listeners, and strong ordering of events matching multiple filters.

Reviewer: `topic` isn't new - this paragraph makes it sound like it's the new thing

> Before this change, each contract listener would only support listening to one specific event from an interface previously defined. Each listener would be comprised of:

Reviewer: ```suggestion
### Single filter format

In versions prior to v1.3.1 listener would only support listening to one specific event.

Each listener would be comprised of:
```

For backwards compatibility, this format is still supported by the API.

### Signature strings

Each filter is identified by a generated `signature` that matches a single event, and each contract listener is identified by a `signature` computed from its filters.

Ethereum provides a string standard for event signatures, of the form `EventName(uint256,bytes)`. Prior to v1.3.1, the signature of each Ethereum contract listener would exactly follow this Ethereum format.

As of v1.3.1, Ethereum signature strings have been changed, because this format does not fully describe the event - particularly because each top-level parameter can in the ABI definition be marked as `indexed`. For example, while the following two Solidity events have the same signature, they are serialized differently due to the different placement of `indexed` parameters, and thus a listener must define both individually to be able to process them:

- ERC-20 `Transfer`

  ```solidity
  event Transfer(address indexed _from, address indexed _to, uint256 _value)
  ```

- ERC-721 `Transfer`

  ```solidity
  event Transfer(address indexed _from, address indexed _to, uint256 indexed _tokenId);
  ```

> The two above will now be expressed in the following manner by the Ethereum blockchain connector:

Reviewer: ```suggestion
The two above are now expressed in the following manner by the Ethereum blockchain connector:
```

```solidity
Transfer(address,address,uint256) [i=0,1]
Transfer(address,address,uint256) [i=0,1,2]
```

The `[i=]` listing at the end of the signature indicates the position of all parameters that are marked as `indexed`.

Building on the blockchain-specific signature format for each event, FireFly will then compute the final signature for each filter and each contract listener as follows:

> - Each filter signature will be a combination of the location and the specific connector event signature, such as `0xa5ea5d0a6b2eaf194716f0cc73981939dca26da1:Changed(address,uint256) [i=0]`

Reviewer: ```suggestion
- Each filter signature is a combination of the location and the specific connector event signature, such as `0xa5ea5d0a6b2eaf194716f0cc73981939dca26da1:Changed(address,uint256) [i=0]`
```

- Each contract listener signature will be a concatenation of all the filter signatures, separated by `;`

### Duplicate filters

> FireFly will restrict the creation of a contract listener containing duplicate filters or superset filters. For example, if two filters are listening to the same event, but one has specified a location and the other hasn't, then the latter will be a superset, and already be listening to all the events matching the first filter. Creation of duplicate or superset filters within a single listener will be blocked.

Reviewer: ```suggestion
FireFly restricts the creation of a contract listener containing duplicate filters.

This includes the special case where one filter is a superset of another filter, due to a wildcard location.

For example, if two filters are listening to the same event, but one has specified a location and the other hasn't, then the latter will be a superset, and already be listening to all the events matching the first filter. Creation of duplicate or superset filters within a single listener will be blocked.
```
